bson/src/main/org/bson/BsonDocument.java

```diff
@@ -764,9 +769,59 @@ public String toString() {
         return toJson();
     }
 
+    @Override
+    public BsonDocument clone() {
+        BsonDocument to = new BsonDocument();
+        for (Entry<String, BsonValue> cur : entrySet()) {
+            if (cur.getValue().isDocument()) {
+                to.put(cur.getKey(), cur.getValue().asDocument().clone());
+            } else if (cur.getValue().isArray()) {
+                to.put(cur.getKey(), cur.getValue().asArray().clone());
+            } else if (cur.getValue().isBinary()) {
+                to.put(cur.getKey(), BsonBinary.clone(cur.getValue().asBinary()));
+            } else if (cur.getValue().isJavaScriptWithScope()) {
+                to.put(cur.getKey(), BsonJavaScriptWithScope.clone(cur.getValue().asJavaScriptWithScope()));
+            } else {
+                to.put(cur.getKey(), cur.getValue());
+            }
+        }
+        return to;
+    }
+
     private void throwIfKeyAbsent(final Object key) {
         if (!containsKey(key)) {
             throw new BsonInvalidOperationException("Document does not contain key " + key);
         }
     }
+
+    private Object writeReplace() {
```

Comment:
This seems magic as its usecase isn't obvious.  

Suggestion link to the Spec in the javadoc https://docs.oracle.com/javase/6/docs/platform/serialization/spec/output.html 

Might make it more maintainable in the future.